contracts/src/v0.8/functions/v1_3_0_zksync/FunctionsBilling.sol

```diff
@@ -0,0 +1,440 @@
+  function getUsdPerUnitLink() public view returns (uint256, uint8) {
+    (, int256 usdPerUnitLink, , uint256 timestamp, ) = s_linkToUsdFeed.latestRoundData();
+    // solhint-disable-next-line not-rely-on-time
+    if (s_config.feedStalenessSeconds < block.timestamp - timestamp && s_config.feedStalenessSeconds > 0) {
+      return (s_config.fallbackUsdPerUnitLink, s_config.fallbackUsdPerUnitLinkDecimals);
+    }
+    if (usdPerUnitLink <= 0) {
+      revert InvalidUsdLinkPrice(usdPerUnitLink);
+    }
+    return (uint256(usdPerUnitLink), s_linkToUsdFeed.decimals());
+  }
+
+  function _getJuelsFromUsd(uint256 amountUsd) private view returns (uint96) {
+    (uint256 usdPerLink, uint8 decimals) = getUsdPerUnitLink();
+    // (usd) * (10**18 juels/link) * (10**decimals) / (link / usd) = juels
+    // There are only 1e9*1e18 = 1e27 juels in existence, should not exceed uint96 (2^96 ~ 7e28)
+    return SafeCast.toUint96((amountUsd * 10 ** (18 + decimals)) / usdPerLink);
+  }
+
+  // ================================================================
+  // |                       Cost Estimation                        |
+  // ================================================================
+
+  /// @inheritdoc IFunctionsBilling
+  function estimateCost(
+    uint64 subscriptionId,
+    bytes calldata data,
+    uint32 callbackGasLimit,
+    uint256 gasPriceWei
+  ) external view override returns (uint96) {
+    _getRouter().isValidCallbackGasLimit(subscriptionId, callbackGasLimit);
+    // Reasonable ceilings to prevent integer overflows
+    if (gasPriceWei > REASONABLE_GAS_PRICE_CEILING) {
+      revert InvalidCalldata();
+    }
+    uint72 adminFee = getAdminFeeJuels();
+    uint72 donFee = getDONFeeJuels(data);
+    uint72 operationFee = getOperationFeeJuels();
+    return _calculateCostEstimate(callbackGasLimit, gasPriceWei, donFee, adminFee, operationFee);
+  }
+
+  /// @notice Estimate the cost in Juels of LINK
+  // that will be charged to a subscription to fulfill a Functions request
+  // Gas Price can be overestimated to account for flucuations between request and response time
+  function _calculateCostEstimate(
+    uint32 callbackGasLimit,
+    uint256 gasPriceWei,
+    uint72 donFeeJuels,
+    uint72 adminFeeJuels,
+    uint72 operationFeeJuels
+  ) internal view returns (uint96) {
+    // If gas price is less than the minimum fulfillment gas price, override to using the minimum
+    if (gasPriceWei < s_config.minimumEstimateGasPriceWei) {
+      gasPriceWei = s_config.minimumEstimateGasPriceWei;
+    }
+
+    uint256 gasPriceWithOverestimation = gasPriceWei +
+      ((gasPriceWei * s_config.fulfillmentGasPriceOverEstimationBP) / 10_000);
+    /// @NOTE: Basis Points are 1/100th of 1%, divide by 10_000 to bring back to original units
+
+    uint256 executionGas = s_config.gasOverheadBeforeCallback + s_config.gasOverheadAfterCallback + callbackGasLimit;
+    uint256 l1FeeWei = ZKSyncChainSpecificUtil._getCurrentTxL1GasFees();
```

Comment:
i see 2 use cases of `_getCurrentTxL1GasFees` to get L1 fee. i am not sure if this will give the correct value.
bc we will not know the exact gas cost for updating storage until the entire transaction is executed.
i have this [document](https://smartcontract-it.atlassian.net/wiki/spaces/DEVSVCS/pages/932741264/ZKSync+Automation+Registry+Audit+Document) for auditors and this section explains certain concepts.

traditionally, automation estimates the gas cost by (the upkeep gas limit + L1 gas cost). however, we cannot reliably estimate the L1 gas cost due to the reason above. we have to forward all the gas to gas bound caller to bound the gas usage. this means the max gas usage will simply be the gas limit. so you can see the l1 cost [here](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/automation/v2_3_zksync/ZKSyncAutomationRegistryBase2_3.sol#L758) is 0.
if functions were to follow the similar path and use gas bound caller with callback gas limit, we need to basically provide the gas cost estimation based on the gas limit for the function callback and use 0 for l1 gas cost.